test/base/src/androidTest/kotlin/io/realm/test/shared/QueryTests.kt

```diff
@@ -0,0 +1,779 @@
+            }
+
+        // Conflicting descriptors, query string vs function - Bob, Ruth, Stacy
+        realm.query(Sample::class)
+            .query("intField > 2 SORT(stringField ASCENDING)")
+            .sort(Sample::stringField.name, Sort.DESCENDING)
+            .find()
+            .let { results ->
+                assertEquals(3, results.size)
+                assertEquals(bob, results[2].stringField)
+                assertEquals(ruth, results[1].stringField)
+                assertEquals(stacy, results[0].stringField)
+            }
+
+        // Invalid descriptor in query string
+        assertFailsWith<IllegalArgumentException> {
+            realm.query(Sample::class)
+                .query("intField > 2 SORT(stringField DESCENDINGGGGGGGGGG)")
+        }
+    }
+
+    @Test
+    fun composedQuery_withDescriptorAndQueryAgain() {
+        val value1 = 1
+        val value2 = 2
+        val value3 = 3
+        val value4 = 4
+        val sylvia = "Sylvia"
+        val stacy = "Stacy"
+        val ruth = "Ruth"
+
+        realm.writeBlocking {
+            copyToRealm(
+                Sample().apply {
+                    intField = value1
+                    stringField = sylvia
+                }
+            )
+            copyToRealm(
+                Sample().apply {
+                    intField = value2
+                    stringField = sylvia // intentionally repeated
+                }
+            )
+            copyToRealm(
+                Sample().apply {
+                    intField = value3
+                    stringField = stacy
+                }
+            )
+            copyToRealm(
+                Sample().apply {
+                    intField = value4
+                    stringField = ruth
+                }
+            )
+        }
+
+        // Ruth 4, Stacy 3
+        realm.query(Sample::class)
+            .distinct(Sample::stringField.name) // Sylvia, Stacy, Ruth
+            .sort(Sample::stringField.name, Sort.ASCENDING) // Ruth, Stacy, Sylvia
+            .limit(2) // Ruth 4, Stacy 3
+            .find { results ->
+                assertEquals(2, results.size)
+                assertEquals(ruth, results[0].stringField)
+                assertEquals(stacy, results[1].stringField)
+            }
+
+        // Stacy 3, Sylvia 1
+        // One would expect this to be Stacy 3 but notice the last 'query', it moves the descriptors
+        // to the end - this is how it is also implemented in Java
+        realm.query(Sample::class)
+            .distinct(Sample::stringField.name) // normal execution: Sylvia 1, Stacy 3, Ruth 4
+            .sort(
+                Sample::stringField.name,
+                Sort.ASCENDING
+            ) // normal execution: Ruth 4, Stacy 4, Sylvia 1
+            .limit(2) // normal execution: Ruth 4, Stacy 3
+            .query("intField < $value4") // this puts all the previous descriptors at the end of this query!
```

Comment:
We could get around this by introducing yet another layer with a RealmQueryDescriptor interface, but probably a bit overkill? At least I have not heard any complaints about this behaviour in Java.

Reply:
My first implementation used a descriptor that kept track of the different elements in the query string. Then I found out that in Java we build queries with descriptors using raw predicates, which in essence does what we are doing here. I think we should avoid changing the semantics to preserve consistency across the two SDKs...

... though I personally find it slightly annoying that descriptors are moved to the end in queries (they do not do that in results - that's one of the reasons why James suggested we used `realm_results_t` and not `realm_query_t` for queries on `RealmResults`) but I see and understand why it's done like that. Again, I would leave it like this for consistency.